compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -2010,7 +2010,18 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             {
                 if let ObligationCauseCode::Pattern { span: Some(span), .. } = *cause.code() {
                     if let Ok(snippet) = self.tcx.sess.source_map().span_to_snippet(span) {
-                        let suggestion = if expected_def.is_struct() {
+                        let suggestion = if exp_found.found.is_unit() {
+                            let exp_str = format!("{:?}", expected_def);
+
+                            // `RangeInclusive`'s fields are private, we have to call methods instead.
+                            if exp_str.contains("core::ops::RangeInclusive")
+                                || exp_str.contains("std::ops::RangeInclusive")
+                            {
```

Comment:
I don't think this is the right approach. I would instead add a `&hir::Expr` (or a `HirId` to get the expression back) to `Pattern` so that we can then call `.precedence()` and compare it to `ExprPrecedence::Field`. This will work regardless of the types. Also, with access to the expression itself, it'd be nice to see if notice these kind of range literals to suggest changing the range to the start or end, instead of constructing the range and then field accessing one of those.

Reply:
Thanks, I've tried to implement your suggestion but it would be a bigger change than I thought. I'm going to ask some questions on Zulip once I get some time to summarize them!

Reply:
Switching to waiting on author then :) Feel free to request a review when ready, thanks!

@rustbot author